File: app/src/main/java/com/syncplay/PlaybackManager.java
package com.syncplay;

import java.util.ArrayList;

/**
 * Created by vishal on 10/14/16.
 */

class PlaybackManager {
    static PlaybackState globalPlaybackState = PlaybackState.UNKNOWN;

    static void play() {
        for (PlaybackDevice device:
                DeviceManager.devices) {
            if (device.getState() != PlaybackState.ERROR && device.getState() != PlaybackState.DROP) {
                device.play();
            }
        }

        PlaybackManager.globalPlaybackState = PlaybackState.PLAY;
    }

    static void pause() {
        for (PlaybackDevice device :
                DeviceManager.devices) {
            if (device.getState() == PlaybackState.PLAY) {
                device.pause();
            }
        }

        PlaybackManager.globalPlaybackState = PlaybackState.PAUSE;
    }

    static void stop() {
        for (PlaybackDevice device :
                DeviceManager.devices) {
            if (device.getState() == PlaybackState.PLAY || device.getState() == PlaybackState.PAUSE) {
                device.stop();
            }
        }

        PlaybackManager.globalPlaybackState = PlaybackState.STOP;
    }

    static void smartPlay() {
        if (PlaybackManager.globalPlaybackState == PlaybackState.UNKNOWN
                || PlaybackManager.globalPlaybackState == PlaybackState.READY
                || PlaybackManager.globalPlaybackState == PlaybackState.PAUSE
                || PlaybackManager.globalPlaybackState == PlaybackState.STOP) {
            play();
        } else {
            pause();
        }
    }
}

File: app/src/main/java/com/syncplay/PlaybackManager.kt
package com.syncplay

internal object PlaybackManager {
    var globalPlaybackState = PlaybackState.UNKNOWN

    fun play() {
        DeviceManager.devices
                .filter { it.state != PlaybackState.ERROR && it.state != PlaybackState.DROP }
                .forEach { it.play() }

        PlaybackManager.globalPlaybackState = PlaybackState.PLAY
    }

    fun pause() {
        DeviceManager.devices
                .filter { it.state == PlaybackState.PLAY }
                .forEach { it.pause() }

        PlaybackManager.globalPlaybackState = PlaybackState.PAUSE
    }

    fun stop() {
        DeviceManager.devices
                .filter { it.state == PlaybackState.PLAY || it.state == PlaybackState.PAUSE }
                .forEach { it.stop() }

        PlaybackManager.globalPlaybackState = PlaybackState.STOP
    }

    fun smartPlay() {
        if (PlaybackManager.globalPlaybackState == PlaybackState.UNKNOWN
                || PlaybackManager.globalPlaybackState == PlaybackState.READY
                || PlaybackManager.globalPlaybackState == PlaybackState.PAUSE
                || PlaybackManager.globalPlaybackState == PlaybackState.STOP) {
            play()
        } else {
            pause()
        }
    }
}
